DocPrepare/PrepareDocs.py

```diff
@@ -0,0 +1,137 @@
+
+        if s3_bucket_name:
+            self.s3_bucket = S3Connection().get_bucket(s3_bucket_name)
+        else:
+            self.s3_bucket = None
+
+    def parse_date(self, date):
+        """ Given string like this: 2013-03-20T17:11:17Z returns date in
+        format YYYY-MM-DD"""
+
+        if date:
+            return date.split('T')[0]
+
+    def parse_tika_metadata(self, metadata_file, metadata):
+        """ Parses metadata from created by Tika """
+
+        with open(os.path.join(metadata_file), 'r') as f:
+            try:
+                tika_metadata = json.loads(f.read())
+            except:
+                tika_metadata = {}
+
+            if not metadata.get('file_type'):
+                metadata['file_type'] = tika_metadata.get(
+                    ['dc:format'], '').replace('application/', '')
+            if not metadata.get('date_released'):
+                metadata['date_released'] = self.parse_date(
+                    tika_metadata.get('Last-Save-Date'))
+            if not metadata.get('title'):
+                metadata['title'] = tika_metadata.get('title')
+
+            metadata['pages'] = tika_metadata.get('xmpTPg:NPages')
+            metadata['date_created'] = self.parse_date(
+                tika_metadata.get('meta:creation-date'))
+
+            return metadata
+
+    def prepare_metadata(self, root, base_file):
+        """ Prepares metadata from Tika metadata file, the file location,
+        and unique parser if available """
+
+        if self.unique_parser:
+            metadata = self.unique_parser(
+                os.path.join(root, base_file + ".json"))
+        else:
+            metadata = {}
+
+        metadata = self.parse_tika_metadata(
+            os.path.join(root, base_file + "_metadata.json"),
+            metadata=metadata)
+
+        return metadata
+
+    def prepare_file_location(self, metadata, root, base_file):
+        """ Adds file location to metadata """
+
+        if self.s3_bucket:
+
+            # Upload pdf
+            doc_location = os.path.join(
+                root, base_file + "." + metadata['file_type'])
+
+            k = Key(self.s3_bucket)
+            k.key = doc_location
+            k.set_contents_from_filename(doc_location)
+
+            # Upload text
+            text_location = os.path.join(root, base_file + ".txt")
+            if os.path.exists(text_location):
+                k = Key(self.s3_bucket)
+                k.key = text_location
+                k.set_contents_from_filename(text_location)
+
+        else:
+            f_dir = os.path.split(root)[-1]
+            doc_location = os.path.join(
+                f_dir, base_file + "." + metadata['file_type'])
+            metadata['doc_location'] = doc_location
+
+    def write_manifest(self, manifest, directory_path):
+        """ Writes manifest files """
+
+        manifest_location = os.path.join(directory_path, "manifest.yaml")
+
+        if self.s3_bucket:
```

Comment:
It might make more sense to abstract this into a write-file method that's only concerned with writing out a file. That way, the S3 logic is only in one place and not peppered throughout the rest of the code.